wagtail/contrib/redirects/views.py

```diff
@@ -36,65 +36,91 @@
-    except InvalidPage:
-        raise Http404
 
-    # Render template
-    if request.headers.get("x-requested-with") == "XMLHttpRequest":
-        return TemplateResponse(
-            request,
-            "wagtailredirects/results.html",
-            {
-                "ordering": ordering,
-                "redirects": redirects,
-                "query_string": query_string,
-            },
-        )
-    else:
-        return TemplateResponse(
-            request,
-            "wagtailredirects/index.html",
-            {
-                "ordering": ordering,
-                "redirects": redirects,
-                "query_string": query_string,
-                "search_form": SearchForm(
-                    data={"q": query_string} if query_string else None,
-                ),
-                "user_can_add": permission_policy.user_has_permission(
-                    request.user, "add"
-                ),
-            },
-        )
+class Index(IndexView):
+    template_name = "wagtailredirects/index.html"
+    results_template_name = "wagtailredirects/results.html"
+    any_permission_required = ["add", "change", "delete"]
+    permission_policy = permission_policy
+    model = Redirect
+    header_icon = "redirects"
+    add_item_label = gettext_lazy("Add redirect")
+    context_object_name = "redirects"
+    index_url_name = "wagtailredirects:index"
+    add_url_name = "wagtailredirects:add"
+    edit_url_name = "wagtailredirects:edit"
+    delete_url_name = "wagtailredirects:delete"
+    index_results_url_name = "wagtailredirects:index_results"
+    default_ordering = "old_path"
+    paginate_by = 20
+    is_searchable = True
+    page_title = gettext_lazy("Redirects")
+    columns = [
+        TitleColumn(
+            "old_path",
+            label=gettext_lazy("From"),
+            width="35%",
+            url_name="wagtailredirects:edit",
+            sort_key="old_path",
+        ),
+        Column(
+            "site",
+            label=gettext_lazy("Site"),
+            width="30%",
+        ),
+        RedirectPageColumn(
+            "redirect_page",
+            label=gettext_lazy("To"),
+            width="15%",
+        ),
+        TypeColumn(
+            "get_is_permanent_display",
+            label=gettext_lazy("Type"),
+            width="15%",
+        ),
+    ]
+
+    def setup(self, request, *args, **kwargs):
+        super().setup(request, *args, **kwargs)
+        self.redirects = Redirect.objects.prefetch_related("redirect_page", "site")
+        self.query_string = request.GET.get("q", "")
+
+    def get_queryset(self):
+        redirects = self.redirects
+        if self.is_searching:
+            if self.query_string:
+                redirects = redirects.filter(
+                    Q(old_path__icontains=self.query_string)
+                    | Q(redirect_page__url_path__icontains=self.query_string)
+                    | Q(redirect_link__icontains=self.query_string)
```

Comment:
This can probably be replaced by setting `search_fields = ["old_path", "redirect_page__url_path", "redirect_link"]` on the view class (not sure if it supports related fields though)